monotonic time jumped back

## Bug Report

<!-- Thanks for your bug report! Don't worry if you can't fill out all the sections. -->

### What version of TiKV are you using?
v6.1.2

### What operating system and CPU are you using?
ubuntu

### Steps to reproduce
Run Jepsen test configured with the append workload and the kill/pause/partition nemesis

### What did you expect?
monotonic time should be always increasing 

### What happened?
We didn't configure any clock changing nemesis, but the monotonic time jumped back. The following is the bug report from the logs:
`./20230316T152153.412Z/n4/kv.log:[2023/03/16 15:22:53.812 +00:00] [ERROR] [time.rs:372] ["monotonic time jumped back, 7968842.265 -> 7968842.265"]`

[2024/03/07 17:59:29.043 +08:00] [ERROR] [time.rs:374] ["monotonic time jumped back, 1581214.068 -> 1581214.057"]
[2024/03/07 17:59:29.043 +08:00] [ERROR] [time.rs:374] ["monotonic time jumped back, 1581214.068 -> 1581214.057"]
[2024/03/07 17:59:29.045 +08:00] [FATAL] [lib.rs:510] ["elapsed=56976988; when=56976976"] [backtrace="   0: tikv_util::set_panic_hook::{{closure}}\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tikv_util/src/lib.rs:509:18\n   1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2032:9\n      std::panicking::rust_panic_with_hook\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:692:13\n   2: std::panicking::begin_panic_handler::{{closure}}\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:579:13\n   3: std::sys_common::backtrace::__rust_end_short_backtrace\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:137:18\n   4: rust_begin_unwind\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:575:5\n   5: core::panicking::panic_fmt\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:65:14\n   6: tokio_timer::wheel::Wheel<T>::poll\n   7: tokio_timer::timer::Timer<T,N>::process\n             at rust/git/checkouts/tokio-8e927faba632ed16/4394380/tokio-timer/src/timer/mod.rs:272:33\n      <tokio_timer::timer::Timer<T,N> as tokio_executor::park::Park>::park\n             at rust/git/checkouts/tokio-8e927faba632ed16/4394380/tokio-timer/src/timer/mod.rs:379:9\n      tokio_timer::timer::Timer<T,N>::turn\n             at rust/git/checkouts/tokio-8e927faba632ed16/4394380/tokio-timer/src/timer/mod.rs:256:21\n   8: tikv_util::timer::start_global_steady_timer::{{closure}}\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tikv_util/src/timer.rs:209:17\n      <std::thread::Builder as tikv_util::sys::thread::StdThreadBuildWrapper>::spawn_wrapper::{{closure}}\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tikv_util/src/sys/thread.rs:427:23\n      std::sys_common::backtrace::__rust_begin_short_backtrace\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:121:18\n   9: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:551:17\n      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panic/unwind_safe.rs:271:9\n      std::panicking::try::do_call\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:483:40\n      std::panicking::try\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:447:19\n      std::panic::catch_unwind\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:137:14\n      std::thread::Builder::spawn_unchecked_::{{closure}}\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:550:30\n      core::ops::function::FnOnce::call_once{{vtable.shim}}\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:513:5\n  10: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2000:9\n      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2000:9\n      std::sys::unix::thread::Thread::new::thread_start\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/unix/thread.rs:108:17\n  11: start_thread\n  12: __GI___clone3\n"] [location=/rust/git/checkouts/tokio-8e927faba632ed16/4394380/tokio-timer/src/wheel/mod.rs:225] [thread_name=steady-timer]

--- a/components/tikv_util/src/timer.rs
+++ b/components/tikv_util/src/timer.rs
@@ -3,10 +3,7 @@
 use std::{
     cmp::{Ord, Ordering, Reverse},
     collections::BinaryHeap,
-    sync::{
-        atomic::{AtomicU64, Ordering as AtomicOrdering},
-        mpsc, Arc,
-    },
+    sync::{mpsc, Arc, Mutex},
     thread::Builder,
     time::Duration,
 };
@@ -200,40 +197,35 @@ fn start_global_steady_timer() -> SteadyTimer {
 ///
 /// [^1]: https://github.com/tokio-rs/tokio/pull/515
 struct RatchetClock<T: Now> {
-    clock: T,
+    // Although the RatchetClock has to be thread-safe (Now requires Sync+Send),
+    // the [`tokio_timer::timer::Clock`] that wraps it can only be accessed by
+    // the timer thread. Therefore, performance penalty of using a Mutex is negligible.
+    state: Mutex<RatchetClockState<T>>,
+}
+
+struct RatchetClockState<T> {
     start: std::time::Instant,
+    clock: T,
 
     // The last recorded time in milliseconds.
     //
     // Using u64 is sufficient here because tokio-timer operates with millisecond
     // precision, and `u64::MAX` milliseconds represent an extremely large time span
     // (several million years).
-    last_now_ms: AtomicU64,
+    last_now_ms: u64,
 }
 
 impl<T: Now> Now for RatchetClock<T> {
     fn now(&self) -> std::time::Instant {
-        let now = self.clock.now();
-        let now_ms = now.saturating_duration_since(self.start).as_millis() as u64;
-        let mut last_now_ms = self.last_now_ms.load(AtomicOrdering::Relaxed);
-        loop {
-            if now_ms > last_now_ms {
-                match self.last_now_ms.compare_exchange_weak(
-                    last_now_ms,
-                    now_ms,
-                    AtomicOrdering::SeqCst,
-                    AtomicOrdering::SeqCst,
-                ) {
-                    Ok(_) => {
-                        return self.start + Duration::from_millis(now_ms);
-                    }
-                    Err(last_now) => {
-                        last_now_ms = last_now;
-                    }
-                }
-            } else {
-                return self.start + Duration::from_millis(last_now_ms);
-            }
+        let mut state = self.state.lock().unwrap();
+        let now = state.clock.now();
+        let now_ms = now.saturating_duration_since(state.start).as_millis() as u64;
+
+        if now_ms > state.last_now_ms {
+            state.last_now_ms = now_ms;
+            state.start + Duration::from_millis(now_ms)
+        } else {
+            state.start + Duration::from_millis(state.last_now_ms)
         }
     }
 }
@@ -241,9 +233,11 @@ impl<T: Now> Now for RatchetClock<T> {
 impl<T: Now> RatchetClock<T> {
     fn new(clock: T) -> Self {
         RatchetClock {
-            clock,
-            start: std::time::Instant::now(),
-            last_now_ms: AtomicU64::new(0),
+            state: Mutex::new(RatchetClockState {
+                clock,
+                start: std::time::Instant::now(),
+                last_now_ms: 0,
+            }),
         }
     }
 }
@@ -271,7 +265,7 @@ fn start_timer_thread<T: Now>(name: &str, clock: T) -> Handle {
 
 #[cfg(test)]
 mod tests {
-    use std::sync::atomic::AtomicBool;
+    use std::sync::atomic::{AtomicBool, Ordering as AtomicOrdering};
 
     use futures::{compat::Future01CompatExt, executor::block_on};
 

--- a/components/tikv_util/src/timer.rs
+++ b/components/tikv_util/src/timer.rs
@@ -3,7 +3,10 @@
 use std::{
     cmp::{Ord, Ordering, Reverse},
     collections::BinaryHeap,
-    sync::{mpsc, Arc},
+    sync::{
+        atomic::{AtomicU64, Ordering as AtomicOrdering},
+        mpsc, Arc,
+    },
     thread::Builder,
     time::Duration,
 };
@@ -92,27 +95,16 @@ impl<T> Ord for TimeoutTask<T> {
     }
 }
 
-lazy_static! {
-    pub static ref GLOBAL_TIMER_HANDLE: Handle = start_global_timer("timer");
+struct SystemClock;
+
+impl Now for SystemClock {
+    fn now(&self) -> std::time::Instant {
+        std::time::Instant::now()
+    }
 }
 
-/// Create a global timer with specific thread name.
-pub fn start_global_timer(name: &str) -> Handle {
-    let (tx, rx) = mpsc::channel();
-    let props = crate::thread_group::current_properties();
-    Builder::new()
-        .name(thd_name!(name))
-        .spawn_wrapper(move || {
-            crate::thread_group::set_properties(props);
-
-            let mut timer = tokio_timer::Timer::default();
-            tx.send(timer.handle()).unwrap();
-            loop {
-                timer.turn(None).unwrap();
-            }
-        })
-        .unwrap();
-    rx.recv().unwrap()
+lazy_static! {
+    pub static ref GLOBAL_TIMER_HANDLE: Handle = start_timer_thread("timer", SystemClock);
 }
 
 /// A struct that marks the *zero* time.
@@ -196,28 +188,91 @@ impl Default for SteadyTimer {
 }
 
 fn start_global_steady_timer() -> SteadyTimer {
-    let (tx, rx) = mpsc::channel();
     let clock = SteadyClock::default();
-    let clock_ = clock.clone();
+    let handle = start_timer_thread("steady-timer", clock.clone());
+    SteadyTimer { clock, handle }
+}
+
+/// A clock that ratchets forward.
+///
+/// It is used to workaround a panic[^1] in tokio-timer when the clock goes
+/// backward, which is possible in some environments, e.g., aliyun ECS g7.
+///
+/// [^1]: https://github.com/tokio-rs/tokio/pull/515
+struct RatchetClock<T: Now> {
+    clock: T,
+    start: std::time::Instant,
+
+    // The last recorded time in milliseconds.
+    //
+    // Using u64 is sufficient here because tokio-timer operates with millisecond
+    // precision, and `u64::MAX` milliseconds represent an extremely large time span
+    // (several million years).
+    last_now_ms: AtomicU64,
+}
+
+impl<T: Now> Now for RatchetClock<T> {
+    fn now(&self) -> std::time::Instant {
+        let now = self.clock.now();
+        let now_ms = now.saturating_duration_since(self.start).as_millis() as u64;
+        let mut last_now_ms = self.last_now_ms.load(AtomicOrdering::Relaxed);
+        loop {
+            if now_ms > last_now_ms {
+                match self.last_now_ms.compare_exchange_weak(
+                    last_now_ms,
+                    now_ms,
+                    AtomicOrdering::SeqCst,
+                    AtomicOrdering::SeqCst,
+                ) {
+                    Ok(_) => {
+                        return now;
+                    }
+                    Err(last_now) => {
+                        last_now_ms = last_now;
+                    }
+                }
+            } else {
+                return self.start + Duration::from_millis(last_now_ms);
+            }
+        }
+    }
+}
+
+impl<T: Now> RatchetClock<T> {
+    fn new(clock: T) -> Self {
+        RatchetClock {
+            clock,
+            start: std::time::Instant::now(),
+            last_now_ms: AtomicU64::new(0),
+        }
+    }
+}
+
+/// Start a timer thread with specific thread name and clock source.
+fn start_timer_thread<T: Now>(name: &str, clock: T) -> Handle {
+    let (tx, rx) = mpsc::channel();
+    let props = crate::thread_group::current_properties();
+    let ratchet_clock = RatchetClock::new(clock);
+    let clock = Clock::new_with_now(ratchet_clock);
     Builder::new()
-        .name(thd_name!("steady-timer"))
+        .name(thd_name!(name))
         .spawn_wrapper(move || {
-            let c = Clock::new_with_now(clock_);
-            let mut timer = tokio_timer::Timer::new_with_now(ParkThread::new(), c);
+            crate::thread_group::set_properties(props);
+
+            let mut timer = tokio_timer::Timer::new_with_now(ParkThread::new(), clock);
             tx.send(timer.handle()).unwrap();
             loop {
                 timer.turn(None).unwrap();
             }
         })
         .unwrap();
-    SteadyTimer {
-        clock,
-        handle: rx.recv().unwrap(),
-    }
+    rx.recv().unwrap()
 }
 
 #[cfg(test)]
 mod tests {
+    use std::sync::atomic::AtomicBool;
+
     use futures::{compat::Future01CompatExt, executor::block_on};
 
     use super::*;
@@ -275,5 +330,38 @@ mod tests {
             end,
             elapsed
         );
+    }
+
+    #[test]
+    fn test_ratchet_clock() {
+        struct BadClock {
+            backward: AtomicBool,
+        }
+        impl Now for BadClock {
+            fn now(&self) -> std::time::Instant {
+                let now = std::time::Instant::now();
+                if self.backward.load(AtomicOrdering::Relaxed) {
+                    self.backward.store(false, AtomicOrdering::Relaxed);
+                    now - Duration::from_millis(10)
+                } else {
+                    self.backward.store(true, AtomicOrdering::Relaxed);
+                    now
+                }
+            }
+        }
+        let clock = BadClock {
+            backward: AtomicBool::new(false),
+        };
+        let handle = start_timer_thread("timer", clock);
+
+        for i in 0..100 {
+            let deadline = if i % 2 == 0 {
+                std::time::Instant::now() + Duration::from_millis(i)
+            } else {
+                std::time::Instant::now() - Duration::from_millis(i)
+            };
+            let delay = handle.delay(deadline);
+            block_on(delay.compat()).unwrap();
+        }
     }
 }

--- a/components/tikv_util/src/timer.rs
+++ b/components/tikv_util/src/timer.rs
@@ -225,7 +225,7 @@ impl<T: Now> Now for RatchetClock<T> {
                     AtomicOrdering::SeqCst,
                 ) {
                     Ok(_) => {
-                        return now;
+                        return self.start + Duration::from_millis(now_ms);
                     }
                     Err(last_now) => {
                         last_now_ms = last_now;

